### PR TITLE
syncing 

### PR DESCRIPTION
**Pull request by jadberjawi**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds broad experimental support for 4-channel RGB+IR workflows, including `.npy` image handling, new fusion modules, and core model pipeline changes across training, validation, and inference. 🚀

### 📊 Key Changes
- Added multiple new multimodal fusion modules under `ultralytics/crossAttn/`, `ultralytics/mefa/`, and `ultralytics/rsr/` for RGB+IR preprocessing and feature fusion.
- Updated data loading and verification to recognize `.npy` files as image inputs, with validation logic expecting 4-channel arrays.
- Modified preprocessing, warmup, validation, plotting, and prediction paths to handle 4-channel tensors instead of standard 3-channel images.
- Changed `yolo11.yaml` to include `ch: 3`, while model construction in `tasks.py` now injects custom RSR and MEFA modules ahead of the parsed model.
- Adjusted augmentation padding, source handling, and cache logic to support `.npy` datasets and preserve nonstandard channels.
- Introduced several debug prints and hard-coded behaviors, including custom source directory expansion, AMP asset swap to `bus.npy`, and scale selection changes in model parsing.
- Updated package version from `8.3.75` to `8.3.7555`. 🛠️

### 🎯 Purpose & Impact
- Enables experimentation with multispectral or RGB+IR datasets using custom 4-channel inputs and early fusion strategies.
- Expands dataset compatibility by allowing `.npy` files to flow through loading, caching, verification, inference, and visualization.
- Could benefit specialized detection research workflows, especially custom sensor pipelines. 🔬
- Carries high integration risk because it alters core behavior in many shared codepaths, adds debug output, and introduces hard-coded assumptions that may affect standard Ultralytics usage.
- Likely needs tighter scoping, tests, and documentation before merge to ensure compatibility with existing YOLO11 and YOLO26 workflows.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @jadberjawi, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated welcome message to help with review readiness, and an Ultralytics engineer will assist shortly. Please review the checklist below 😊

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Comment by glenn-jocher** [comment]
@jadberjawi closing as appears opened by accident

File: ultralytics/mefa/mefa_ca.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+        # Upscale Attention Maps to Match Input Size
+        ga_rgb = F.interpolate(ga_rgb, size=(x.shape[2], x.shape[3]), mode="bilinear", align_corners=False)
+        ga_ir = F.interpolate(ga_ir, size=(x.shape[2], x.shape[3]), mode="bilinear", align_corners=False)
+
+        # Pass Through Final Inception Block
+        refined_features = self.inception_fused(fused_features)
+
+        # Apply Global Attention
+        final_rgb = refined_features * ga_rgb + refined_features
```
Comment (UltralyticsAssistant) [inline ultralytics/mefa/mefa_ca.py:232]: ❗ **CRITICAL**: `refined_features` comes from `InceptionBlockA(in_channels=256)` and has 256 channels, but `ga_rgb`/`ga_ir` are produced by `GlobalAttention(in_channels=128)` and still have 128 channels after interpolation. Multiplying them here will raise a channel-size mismatch at runtime.

File: ultralytics/data/base.py [hunk 2]
```diff
@@ -152,13 +152,15 @@ def load_image(self, i, rect_mode=True):
...
-                    im = np.load(fn)
+                    im = np.load(f)  # Load as NumPy array
+                    if im.shape[-1] != 4:  # Ensure it's 4-channel
+                        raise ValueError(f"Invalid `.npy` format: {f} has shape {im.shape}")
                 except Exception as e:
-                    LOGGER.warning(f"{self.prefix}WARNING ⚠️ Removing corrupt *.npy image file {fn} due to: {e}")
-                    Path(fn).unlink(missing_ok=True)
-                    im = cv2.imread(f)  # BGR
+                    LOGGER.warning(f"{self.prefix}WARNING ⚠️ Removing corrupt `.npy` file {f} due to: {e}")
+                    Path(f).unlink(missing_ok=True)
```
Comment (UltralyticsAssistant) [inline ultralytics/data/base.py:162]: ⚠️ **HIGH**: On load failure this deletes `f`, which is now the original dataset file path, not the cache file. A transient read/shape error will remove the user's source image from disk. The previous code only removed the cached `.npy` artifact.

File: ultralytics/data/loaders.py [hunk 3]
```diff
@@ -589,7 +589,9 @@ def autocast_list(source):
     files = []
     for im in source:
         if isinstance(im, (str, Path)):  # filename or uri
-            files.append(Image.open(requests.get(im, stream=True).raw if str(im).startswith("http") else im))
+            # MODOFICATION HERE: SWITCHED THE COMMENTED LINE INTO THE ONE BELOW IT
+            # files.append(Image.open(requests.get(im, stream=True).raw if str(im).startswith("http") else im))
+            files.append(im)
```
Comment (UltralyticsAssistant) [inline ultralytics/data/loaders.py:594]: ⚠️ **HIGH**: `autocast_list()` is expected to normalize path inputs into loaded images. Appending the raw path here changes the return type for string/Path inputs and will break downstream preprocessing for common Python API calls like `model(["img.jpg"])` or URL inputs, which previously received PIL images.

File: ultralytics/engine/model.py [hunk 4]
```diff
@@ -555,6 +557,27 @@ def predict(
...
                 self.predictor.save_dir = get_save_dir(self.predictor.args)
+
+        # ✅ MODIFICATIONS HERE: What was happening is that source contains location of the directory where the images are stored and not the location of the image directly 
+        image_files = []
+        if isinstance(source, str) and os.path.isdir(source):
+            image_files = sorted(glob.glob(os.path.join(source, "*.npy")))  # Get all .npy images
+            
+            # Debugging: Ensure images are found
+            if not image_files:
+                raise ValueError(f"🚨 No .npy images found in {source}!")
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/model.py:568]: ⚠️ **HIGH**: This changes directory prediction semantics for all callers: if `source` is a directory without `.npy` files, prediction now raises instead of letting the normal loader handle JPG/PNG contents. That regresses the existing `model.predict("dir/")` path for standard image folders.

File: ultralytics/engine/predictor.py [hunk 5]
```diff
@@ -238,7 +242,7 @@ def stream_inference(self, source=None, model=None, *args, **kwargs):
 
             # Warmup model
             if not self.done_warmup:
-                self.model.warmup(imgsz=(1 if self.model.pt or self.model.triton else self.dataset.bs, 3, *self.imgsz))
+                self.model.warmup(imgsz=(1 if self.model.pt or self.model.triton else self.dataset.bs, 4, *self.imgsz))
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/predictor.py:245]: ⚠️ **HIGH**: This hard-codes warmup to 4 input channels for every model. Standard 3-channel checkpoints will now fail during prediction warmup before inference even starts. The channel count here needs to stay model-dependent rather than globally forced to 4.

File: ultralytics/engine/validator.py [hunk 6]
```diff
@@ -158,7 +158,7 @@ def __call__(self, trainer=None, model=None):
             self.dataloader = self.dataloader or self.get_dataloader(self.data.get(self.args.split), self.args.batch)
 
             model.eval()
-            model.warmup(imgsz=(1 if pt else self.args.batch, 3, imgsz, imgsz))  # warmup
+            model.warmup(imgsz=(1 if pt else self.args.batch, 4, imgsz, imgsz))  # warmup
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/validator.py:161]: ⚠️ **HIGH**: This makes validation warmup unconditional 4-channel input. That breaks existing 3-channel models during val/test runs, even when the dataset and checkpoint are unchanged. Warmup should use the model's actual expected input channels.

File: ultralytics/nn/tasks.py [hunk 7]
```diff
@@ -933,7 +981,7 @@ def parse_model(d, ch, verbose=True):  # model_dict, input_channels(3)
     if scales:
         scale = d.get("scale")
         if not scale:
-            scale = tuple(scales.keys())[0]
+            scale = tuple(scales.keys())[4] # HERE CHANGE THE SCALE
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/tasks.py:984]: ⚠️ **HIGH**: Indexing the 5th scale key here assumes every model YAML has at least five scales and silently changes the default fallback from the first declared scale to `x`. For YAMLs with fewer entries this raises `IndexError`, and for the rest it changes model construction behavior globally.

**Suggested change:**
```suggestion
            scale = tuple(scales.keys())[0]
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The 4-channel support work is substantial, but there are a few blocking regressions: existing 3-channel predict/validate paths now fail because warmup is hard-coded to 4 channels, directory inference is narrowed to `.npy` only, and the MEFA_CA fusion path has a runtime channel mismatch. I’d address those before merging.

💬 Posted 8 inline comments

File: ultralytics/data/base.py [hunk 8]
```diff
@@ -152,13 +152,15 @@ def load_image(self, i, rect_mode=True):
         """Loads 1 image from dataset index 'i', returns (im, resized hw)."""
         im, f, fn = self.ims[i], self.im_files[i], self.npy_files[i]
         if im is None:  # not cached in RAM
-            if fn.exists():  # load npy
+            if f.endswith('.npy'):  # Load .npy file
```
Comment (UltralyticsAssistant) [inline ultralytics/data/base.py:155]: 💡 **MEDIUM**: This no longer reads from the cached `self.npy_files[i]` for normal JPG/PNG datasets. After this change, disk caching only works when the original source file itself ends with `.npy`, so cached `.npy` files created by `cache_images_to_disk()` are never consumed for standard image datasets.

**Comment by jadberjawi** [comment]
Ow sorry I did this by mistake, I wanted to do the oposite, DISCARD

**Comment by glenn-jocher** [comment]
No worries, it’s closed now—if you want to revisit this later, please open a new PR with a smaller scoped change.
